### PR TITLE
Live dashboard: Created current ride section

### PR DESCRIPTION
**Pull request by dawiddawidowski**
Created 4 files to develop the dashboard:
- `database.py` contains utility functions to interact with the RDS database.
    - Establishes database connection
    - Fetches current rider data
    - Fetches personal best data for the current rider

- `utilities.py` contains utility functions to transform data fetched from SQL for use in visualisations.
    - Concatenates rider first and last names
    - Calculates rider age
    - Calculates minimum/maximum heart rate thresholds for the current rider
    - Verifies whether current heart rate is within the thresholds
    
- `visualisations.py` contains functions for visualising Deloton Bike ride data on the Streamlit app.
    - Generates document title
    - Generates current ride header
    - Calculates and generates caption for the last updated time.
    - Generates an expandable header with current rider personal info
    - Generates current ride metrics header
    - Generates warning if heart rate is abnormal
    - Generates an expandable header with current riders personal best metrics
   
- `dashboard.py` main script that calls the functions from the other 3 files.
    - Refreshes the section every set period of time (period set using global constant, currently 20 seconds)
    - Updates last updated time counter every set period of time (period set using global constant, currently 1 second)

Created dashboard wireframe and added to the `diagrams` folder.

Updated README with setup instructions and wireframe images.

### CODE REVIEW

File: dashboard/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+- It is recommended before stating any installations that you make a new virtual environment (`venv`).
+
+- A new environment will be required for each folder in this repository.
+
+- Install all requirements for this folder by running `pip3 install -r requirements.txt`.
+
+- Create a `.env` file by running `touch .env`
+
+- You need a database called `postgres`.
+
```
Comment (CharliedeanCS) [inline dashboard/README.md:16]:
```suggestion
## 🏃 Running the dashboard locally

To run locally use the command `streamlit run dashboard.py`
```

We had this section last week which you could include. Just a suggestion

File: dashboard/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Deleton Bike Analysis Dashboard
```
Comment (CharliedeanCS) [inline dashboard/README.md:1]: README looks good. Nicely done

File: dashboard/dashboard.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+"""
+Dashboard script to establish connection to the RDS database, fetch data using SQL queries and
+create visualisations in a Streamlit app (using functions from the `database.py`,
+`visualisations.py` and `utilities.pyz files as necessary).
+"""
```
Comment (CharliedeanCS) [inline dashboard/dashboard.py:5]: Nicely detailed module docstring

File: dashboard/dashboard.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                      get_current_ride_data, get_current_ride_data_highest)
+from utilities import get_current_rider_name, is_heart_rate_abnormal
+from visualisations import (get_current_ride_header, get_dashboard_title,
+                            get_current_ride_header_personal_info, get_current_ride_metrics,
+                            get_current_ride_personal_best_metrics, get_last_updated_current_ride,
+                            get_heart_rate_warning)
+
+
+CURRENT_RIDE_REFRESH_RATE = 5
+LAST_UPDATED_COUNT_INCREMENT = 1
```
Comment (CharliedeanCS) [inline dashboard/dashboard.py:24]: Nice use of constants

File: dashboard/dashboard.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    conn = get_database_connection()
+
+    while True:
+        # Auto-refresh the current ride section
+        last_updated_placeholder = main_current_ride(conn)
+        update_time = datetime.now()
+        for i in range(CURRENT_RIDE_REFRESH_RATE):
+            get_last_updated_current_ride(
+                update_time, last_updated_placeholder)
+            time.sleep(LAST_UPDATED_COUNT_INCREMENT)
```
Comment (CharliedeanCS) [inline dashboard/dashboard.py:72]: It look like you are refreshing the dashboard while a while loop and time.sleep. Idk if this makes it easier or harder but I think there is an import for streamlit to autorefresh. https://discuss.streamlit.io/t/streamlit-autorefresh/14519
This link goes over it more.

Comment (CharliedeanCS) [reply dashboard/dashboard.py:72]: Could be worthwhile to checkout. It may be hard to insert

File: dashboard/database.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
+"""Utility functions to interact with the RDS database."""
+
+from os import environ
+
+import psycopg2
+from psycopg2 import extensions, OperationalError
+
+
+def get_database_connection() -> extensions.connection:
```
Comment (CharliedeanCS) [inline dashboard/database.py:9]: I had a look through the code and I dont know if you ever close the connection to the database but everytime the while loop restarts I think a new connection to the database is grabbed. If a database connection is never closed we could run out of connections to the RDS

File: dashboard/database.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+        query = """
+        SELECT Ride.rider_id, first_name, last_name, height, weight, gender, birthdate,
+        heart_rate, power, resistance, elapsed_time, start_time
+        FROM Ride
+        JOIN Rider ON Ride.rider_id = Rider.rider_id
+        JOIN Reading ON Ride.ride_id = Reading.ride_id
+        ORDER BY start_time DESC, elapsed_time DESC
+        LIMIT 1;
+        """
```
Comment (CharliedeanCS) [inline dashboard/database.py:39]: I like the seperation of the query to the execute. Makes it very readable

File: dashboard/database.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    ;
+    """
+
+    parameters = (rider_id, )
+
+    db_cur.execute(query, parameters)
+
+    highest_duration = db_cur.fetchone()
+
+    return highest_duration[0] if highest_duration is not None else None
```
Comment (CharliedeanCS) [inline dashboard/database.py:68]: Nicely done with the return if else

File: dashboard/database.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    """Returns the highest historical power of the rider with the given rider_id."""
+
+    query = """
+    SELECT power
+    FROM ride
+    JOIN Rider ON Ride.rider_id = Rider.rider_id
+    JOIN Reading ON Ride.ride_id = Reading.ride_id
+    WHERE Ride.rider_id = %s
+    ORDER BY power DESC
+    LIMIT 1
```
Comment (CharliedeanCS) [inline dashboard/database.py:102]: Nice simple SQL Querys throughout

**Review by CharliedeanCS — COMMENTED** [review]
Start of review

File: dashboard/database.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    ORDER BY resistance DESC
+    LIMIT 1
+    ;
+    """
+
+    db_cur.execute(query, (rider_id, ))
+
+    highest_resistance = db_cur.fetchone()
+
+    return highest_resistance[0] if highest_resistance is not None else None
```
Comment (CharliedeanCS) [inline dashboard/database.py:131]: Is this easily testable?

File: dashboard/utilities.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
+"""Utility functions to transform data fetched from SQL for use in visualisations."""
+
+from datetime import datetime
+
+
+def get_current_rider_name(current_ride: list) -> str:
+    """Returns a string containing the rider first and last name."""
+    first_name = current_ride[1]
+    last_name = current_ride[2]
+    rider_name = f"{first_name} {last_name}"
```
Comment (CharliedeanCS) [inline dashboard/utilities.py:10]: I like the seperation of variables to improve readability. Very nice

File: dashboard/utilities.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    birthdate = user_details[6]
+    age = calculate_age(birthdate)
+    gender = user_details[5]
+
+    if gender == "female":
+        return round(206 - (0.88 * age))
+    if gender == "male" and age < 40:
+        return round(220 - age)
+
+    return round(208 - (0.7 * age))
```
Comment (CharliedeanCS) [inline dashboard/utilities.py:43]: Nicely done.

File: dashboard/utilities.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    return age
+
+
+def calculate_max_heart_rate(user_details: list) -> int:
+    """
+    Returns the maximum heart rate for the given user based on their age and gender.
+
+    'birthdate' and 'gender' are assumed to be cleaned and
+    always as datetime and str types, respectively.
+    """
```
Comment (CharliedeanCS) [inline dashboard/utilities.py:33]: Nice docstrings throughout

File: dashboard/visualisations.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+def get_current_ride_header(rider_name: str) -> None:
+    """Returns a header for the current ride and the rider's name."""
+    st.header(f"CURRENT RIDE: {rider_name}", divider='blue')
+
+
+def get_last_updated_current_ride(last_update_time: datetime,
+                                  last_updated_placeholder: st.empty) -> None:
+    """Returns a caption under the header with the time since the last data update."""
+
+    # time_delta = (current_time-last_update_time).total_seconds()
```
Comment (CharliedeanCS) [inline dashboard/visualisations.py:26]: Is this supposed to be here / should it be commented out

File: dashboard/visualisations.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    last_updated_placeholder.caption(
+        f"Last updated: {time_delta} seconds ago")
+
+
+def get_current_ride_header_personal_info(user_details: list) -> None:
+    """
+    Gets the main header personal_info for the current ride and displays them.
+    """
+    with st.expander('Personal Info ⛛'):
+        # get metrics
```
Comment (CharliedeanCS) [inline dashboard/visualisations.py:40]: Nice use of comments

File: dashboard/visualisations.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        height = user_details[3]
+        weight = user_details[4]
+        gender = user_details[5]
+        gender_emoji = "♂" if gender == "male" else "♀"
```
Comment (CharliedeanCS) [inline dashboard/visualisations.py:44]: Cool using emojis

File: dashboard/visualisations.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+def get_heart_rate_warning(heart_rate: int) -> None:
+    """Displays a warning message if the heart rate is abnormal for the current ride."""
+    if heart_rate == 0:
+        st.warning(
+            """WARNING! HEART RATE IS ZERO: \n
+            PLEASE PLACE HANDS ON HANDLE BARS OR SEEK ASSISTANCE!""", icon="⚠️")
+    else:
+        st.warning(
+            """WARNING! HEART RATE IS OUTSIDE THE HEALTHY RANGE: \n
+            PLEASE SLOW DOWN OR SEEK ASSISTANCE!""", icon="⚠️")
```
Comment (CharliedeanCS) [inline dashboard/visualisations.py:69]: Nice warnings. The heart rate is also zero for the first few log files as the bike is starting up. I think. Therefore you could check if the duration is above 10 seconds before displaying a warning as you know from this point the bike is up and running

File: dashboard/dashboard.py (hunk 5 above)
Comment (dawiddawidowski) [reply dashboard/dashboard.py:72]: Thanks! I have tried to implement it, but I had trouble with also updating the `Last updated` caption every second in the dashboard. But it's something I can look at again if time.sleep poses any problems.

**Review by CharliedeanCS — APPROVED** [review]
Nice changes to the code based on the review 💯

**Pull request merged**